### ISSUE TITLE
Adding generic.glsl to resources.conf

### ISSUE THREAD

**Issue by HaroldReyiz** [mosra/magnum#443]
Hi,

I've been tinkering with shaders in Magnum. I need to be able to cut an icosphere with a plane (that has an arbitrary normal vector). Actually multiple planes. I have an approach in mind which utilizes geometry shader (and later on, transform feedback for optimization: Run the geometry shader only when the geometry is "dirtied" instead of every frame). I thought I could copy the built-in phong shader and add a geometry shader and other stuff to it. I really like how everything fits together with #defines and resource.addSource() etc.

My question:
I'm using Magnum statically built via vcpkg. I wonder how I could incorporate generic.glsl into all this. Currently I'm thinking of copying the latest master's generic.glsl into my shaders folder. Or I could simply addSource( contents_of_generic_glsl ) in my shader's constructor. But these feel wrong.

Is there a preferred way to do this ? As said in the docs, I want my shader to be compatible with the built-in shaders for debugging purposes.

**Comment by HaroldReyiz** [comment]
By the way is it okay that I'm using Issues in a forum-like fashion ? This is post does not pose an issue for example. Is there a Discord or some other platform that I should be using instead of this ?

**Comment by mosra** [comment]
Hi! I see you discovered the [Gitter chat](https://gitter.im/mosra/magnum) in the meanwhile, but since this question pops up every once in a while, I'll answer it here so it's easier to find for whoever comes across the same problem next.

If you use anything from the Shaders library in your code, it should be enough to request it through shader resources:

```cpp
Utility::Resource rs{"MagnumShaders"};

// …

vert.addSource(rs.get("generic.glsl"));
```

However I would discourage this approach because this thing is intentionally kept internal and thus it could get renamed or disappear at any point. Instead, what I would suggest is propagating the input/output locations from the public shader API. For example:

```cpp
vert.addSource(Utility::formatString(
    "#define POSITION_ATTRIBUTE_LOCATION {}\n"
    "#define NORMAL_ATTRIBUTE_LOCATION {}\n",
    Shaders::Generic3D::Position::Location,
    Shaders::Generic3D::Normal::Location));
```

The above is now included in the [Shaders::Generic docs](https://doc.magnum.graphics/magnum/structMagnum_1_1Shaders_1_1Generic.html#Shaders-Generic-custom) as well.

> I need to be able to cut an icosphere with a plane (that has an arbitrary normal vector)

For this you might be interested in hardware clipping planes. As far as I understand, those are directly in the hardware because CAD applications need them a lot, and so should hopefully be faster than anything performed via a GS. Support for this was added recently (mainly the [GL::Renderer::Feature::ClipDistanceN](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1Renderer.html#aac87c961eb53921e17d66bbf146828aead2634099fd47d8495441e48d59e4e296) enums) and this is a good tutorial: https://prideout.net/clip-planes

**Comment by HaroldReyiz** [comment]
> Instead, what I would suggest is propagating the input/output locations from the public shader API.

That makes sense. I'll use that way, sorry I did not see it in the docs :/

> For this you might be interested in hardware clipping planes.

That seems neat I did not know that. 

Why I wanted to use a geometry shader in the first place is that I need to "stitch" the created hole somehow, i.e. I need a cap to close that hole so that it looks like the surface is milled by a cnc machine (Actually that's exactly what I'm trying to simulate: A sphere, milled by a cnc machine). 
But I was also planning on making the milled surfaces selectable later on. So drawing them as separate entities makes more sense than the whole milled sphere being only 1 object.

But I'll still try the geometry shader approach just to see the GS workflow work.
After I fail miserably at this GS approach, I'll use hardware clipping planes along with separate circle caps to close the gaps.

Thanks a lot for both solutions!

And now I can use use gitter for small quality of life stuff such as this!
Have a nice one.
